File: src/buffer.c
#include <stdio.h>
#include <string.h>

#include "include/buffer.h"
#include "include/macros.h"
#include "include/memory.h"

bool buffer_init(buffer_T* buffer) {
  buffer->capacity = 1024;
  buffer->length = 0;
  buffer->value = nullable_safe_malloc(buffer->capacity * sizeof(char));

  if (!buffer->value) {
    fprintf(stderr, "Error: Failed to initialize buffer with capacity of %zu.\n", buffer->capacity);
    return false;
  }

  buffer->value[0] = '\0';

  return true;
}

buffer_T buffer_new(void) {
  buffer_T buffer;
  buffer_init(&buffer);
  return buffer;
}

char* buffer_value(buffer_T* buffer) {
  return buffer->value;
}

size_t buffer_length(buffer_T* buffer) {
  return buffer->length;
}

size_t buffer_capacity(buffer_T* buffer) {
  return buffer->capacity;
}

size_t buffer_sizeof(void) {
  return sizeof(buffer_T);
}

bool buffer_increase_capacity(buffer_T* buffer, size_t required_length) {
  size_t required_capacity = buffer->length + required_length;

  if (buffer->capacity >= required_capacity) return true;

  size_t new_capacity = required_capacity * 2;
  char* new_value = safe_realloc(buffer->value, new_capacity);

  if (unlikely(new_value == nullptr)) return false;

  buffer->value = new_value;
  buffer->capacity = new_capacity;

  return true;
}

void buffer_append(buffer_T* buffer, const char* text) {
  if (!text || text[0] == '\0') return;

  size_t text_length = strlen(text);

  if (!buffer_increase_capacity(buffer, text_length)) return;

  memcpy(buffer->value + buffer->length, text, text_length);
  buffer->length += text_length;
  buffer->value[buffer->length] = '\0';
}

void buffer_append_char(buffer_T* buffer, char character) {
  if (!buffer_increase_capacity(buffer, 1)) return;

  buffer->value[buffer->length] = character;
  buffer->length++;
  buffer->value[buffer->length] = '\0';
}

void buffer_prepend(buffer_T* buffer, const char* text) {
  if (!text || text[0] == '\0') return;

  size_t text_length = strlen(text);

  if (!buffer_increase_capacity(buffer, text_length)) return;

  memmove(buffer->value + text_length, buffer->value, buffer->length + 1);
  memcpy(buffer->value, text, text_length);

  buffer->length += text_length;
}

void buffer_concat(buffer_T* destination, buffer_T* source) {
  if (source->length == 0) return;
  if (!buffer_increase_capacity(destination, source->length)) return;

  memcpy(destination->value + destination->length, source->value, source->length);
  destination->length += source->length;
  destination->value[destination->length] = '\0';
}

bool buffer_reserve(buffer_T* buffer, size_t min_capacity) {
  size_t required_length = min_capacity - buffer->length;

  return buffer_increase_capacity(buffer, required_length);
}

void buffer_clear(buffer_T* buffer) {
  buffer->length = 0;
  buffer->value[0] = '\0';
}

void buffer_free(buffer_T* buffer) {
  if (!buffer) return;

  free(buffer->value);

  buffer->value = nullptr;
  buffer->length = buffer->capacity = 0;
}

File: rust/src/buffer.rs
#[derive(Debug, Clone)]
pub struct Buffer {
  value: String,
  capacity: usize,
}

impl Buffer {
  pub fn new() -> Self {
    Self {
      value: String::with_capacity(1024),
      capacity: 1024,
    }
  }

  pub fn value(&self) -> &str {
    &self.value
  }

  pub fn length(&self) -> usize {
    self.value.len()
  }

  pub fn capacity(&self) -> usize {
    self.capacity
  }

  pub fn increase_capacity(&mut self, required_length: usize) -> bool {
    let required_capacity = self.length() + required_length;
    if self.capacity >= required_capacity {
      return true;
    }

    let new_capacity = required_capacity * 2;
    self.value.reserve(new_capacity - self.capacity);
    self.capacity = new_capacity;
    true
  }

  pub fn append(&mut self, text: &str) {
    if text.is_empty() {
      return;
    }
    if !self.increase_capacity(text.len()) {
      return;
    }
    self.value.push_str(text);
  }

  pub fn append_char(&mut self, character: char) {
    self.append(&character.to_string());
  }

  pub fn append_repeated(&mut self, character: char, length: usize) {
    if length == 0 {
      return;
    }
    self.append(&character.to_string().repeat(length));
  }

  pub fn append_whitespace(&mut self, length: usize) {
    self.append_repeated(' ', length);
  }

  pub fn prepend(&mut self, text: &str) {
    if text.is_empty() {
      return;
    }
    if !self.increase_capacity(text.len()) {
      return;
    }
    self.value.insert_str(0, text);
  }

  pub fn concat(&mut self, source: &Buffer) {
    if source.length() == 0 {
      return;
    }
    self.append(&source.value);
  }

  pub fn reserve(&mut self, min_capacity: usize) -> bool {
    let required_length = min_capacity.saturating_sub(self.length());
    self.increase_capacity(required_length)
  }

  pub fn clear(&mut self) {
    self.value.clear();
  }
}
